src/Runner.hs

```diff
@@ -59,18 +62,51 @@ instance Semigroup Summary where
     (Summary x1 x2 x3 x4) (Summary y1 y2 y3 y4) = Summary (x1 + y1) (x2 + y2) (x3 + y3) (x4 + y4)
 
 -- | Run all examples from a list of modules.
-runModules :: Bool -> Bool -> Bool -> Interpreter -> [Module [Located DocTest]] -> IO Summary
-runModules fastMode preserveIt verbose repl modules = do
+runModules
+  :: Bool
+  -- ^ Fast mode
+  -> Bool
+  -- ^ Preserve it
+  -> Bool
+  -- ^ Verbose output
+  -> Int
+  -- ^ Number of threads
+  -> Either [String] Interpreter
+  -- ^ Arguments for isolate interpreter or shared interpreter
+  -> [Module [Located DocTest]]
+  -- ^ Modules to test
+  -> IO Summary
+runModules fastMode preserveIt verbose nThreads replE modules = do
   isInteractive <- hIsTerminalDevice stderr
-  ReportState _ _ _ s <- (`execStateT` ReportState 0 isInteractive verbose mempty {sExamples = c}) $ do
-    forM_ modules $ runModule fastMode preserveIt repl
 
+  -- Start a thread pool. It sends status updates to this thread through 'output'.
+  (input, output) <- makeThreadPool nThreads (runModule fastMode preserveIt replE)
+
+  -- Send instructions to threads
+  liftIO (mapM_ (writeChan input) modules)
+
+  let
+    nExamples = (sum . map count) modules
+    initState = ReportState 0 isInteractive verbose mempty {sExamples = nExamples}
+
+  ReportState _ _ _ s <- (`execStateT` initState) $ do
```

Comment:
This actually looks like a misuse of `StateT` to me (instead of `RWST` or even `WriterT`) - but so does master's code. I'll have a look at refactoring.

Reply:
I had a look at this, but there's not that much of a win. In new code I'd work harder to get the monad stack right, but this is working-well-enough code and it doesn't seem to be doing harm.